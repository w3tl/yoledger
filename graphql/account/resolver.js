import { Account as Accounts } from '../../model';

export const Account = {
  id: ({ _id }) => _id,
};

export const Query = {
  async accounts(root, { type }, { userId, connection }) {
    const account = new Accounts(connection.db(), userId);

    return account.find({ type }).toArray();
  },
};

export const Mutation = {
  async addAccount(root, { input }, { userId, connection }) {
    const { name, type, balance } = input;
    const account = new Accounts(connection.db(), userId);
    const { insertedId } = await account.create({
      name,
      balance,
      type,
    });

    return {
      account: await account.findOne({ _id: insertedId }),
    };
  },
};
