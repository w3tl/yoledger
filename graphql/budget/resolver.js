import { Budget as Budgets } from '../../model';

export const Budget = {
  id: ({ _id }) => _id,
  account: ({ account }, _, { dataloaders, userId }) => dataloaders
    .accountByName.load({ name: account, userId }),
};

export const Query = {
  async budget(root, { dateStart, dateEnd, account }, { userId, connection }) {
    const budget = new Budgets(connection.db(), userId);
    const query = {
      $and: [{
        date: { $gte: dateStart },
      }, {
        date: { $lte: dateEnd },
      }],
    };
    if (account) { query.account = account; }

    return budget.find(query).toArray();
  },
};

export const Mutation = {
  async addBudget(root, { input }, { userId, connection }) {
    const { account, date, amount } = input;
    const budget = new Budgets(connection.db(), userId);
    const insertedId = await budget.addBudget({ account, date, amount });

    return {
      budget: await budget.findOne({ _id: insertedId }),
    };
  },
  async updateBudget(root, { id, input: { amount } }, { userId, connection }) {
    const budget = new Budgets(connection.db(), userId);
    await budget.updateBudget(id, { amount });
    return {
      budget: await budget.findOne({ _id: id }),
    };
  },
  async deleteBudget(root, { id }, { userId, connection }) {
    const budget = new Budgets(connection.db(), userId);
    const { deletedCount } = await budget.deleteOne({ _id: id });

    return { success: deletedCount === 1 };
  },
};
