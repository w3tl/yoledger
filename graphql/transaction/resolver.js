import { GraphQLError } from 'graphql';
import { Transaction as Model } from '../../model';

export const Transaction = {
  id: ({ _id }) => _id,
  date: ({ createdAt }) => createdAt,
  source({ source }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ name: source.name, userId });
  },
  destination({ destination }, _, { dataloaders, userId }) {
    return dataloaders.accountByName.load({ name: destination.name, userId });
  },
};

export const Mutation = {
  async addTransaction(root, { input }, { userId, connection }) {
    const {
      source, destination, amount, date: createdAt,
    } = input;
    const transObj = {
      source,
      destination,
      amount,
      createdAt,
    };

    const trans = new Model(connection.db(), userId);
    let _id;
    await connection.withSession(async (session) => {
      _id = await trans.post(transObj, session);
    });

    const transaction = await trans.findOne({ _id });

    return {
      transaction,
    };
  },
  async deleteTransaction(root, { id }, { userId, connection }) {
    const trans = new Model(connection.db(), userId);
    try {
      await connection.withSession(session => trans.unpost(id, session));
    } catch (err) {
      return {
        success: false,
      };
    }

    return {
      success: true,
    };
  },
  async updateTransaction(root, { id, input }, { userId, connection }) {
    const trans = new Model(connection.db(), userId);
    const {
      amount, source, destination, date: createdAt,
    } = input;
    const updates = {
      source,
      destination,
      amount,
      createdAt,
    };
    try {
      let updatedId;
      await connection.withSession(async (session) => {
        updatedId = await trans.update(id, updates, session);
      });
      const transaction = await trans.findOne({ _id: updatedId });
      return { transaction };
    } catch (err) {
      throw new GraphQLError(err.message);
    }
  },
};

export const Query = {
  /* eslint-disable-next-line object-curly-newline */
  async transactions(_, { dateStart, dateEnd, page, itemsPerPage }, { userId, connection }) {
    const trans = new Model(connection.db(), userId);
    const query = {};
    if (dateEnd) {
      query.$and = [
        { createdAt: { $gte: dateStart } },
        { createdAt: { $lte: dateEnd } },
      ];
    } else {
      query.createdAt = { $gte: dateStart };
    }

    return trans.find(query)
      .skip(page * itemsPerPage)
      .limit(itemsPerPage)
      .toArray();
  },
};
