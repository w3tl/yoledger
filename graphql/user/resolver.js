import { User as Model } from '../../model';

export const User = {
  username: ({ _id }) => _id,
};

export const Query = {
  user(root, _, { userId, connection }) {
    const user = new Model(connection.db(), userId);
    return user.findOne();
  },
};
