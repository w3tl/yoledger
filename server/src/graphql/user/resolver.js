export const User = {
  username: ({ _id }) => _id,
};

export const Query = {
  user(root, _, { models: { userModel } }) {
    return userModel.findOne();
  },
};
