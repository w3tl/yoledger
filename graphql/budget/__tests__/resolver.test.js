jest.mock('mongodb');

import { dataloaders } from '../../index';
import { Query, Mutation, Budget } from '../resolver';

const userId = 'admin';

describe('budget resolver', () => {
  const { connection } = require('mongodb');
  const context = {
    dataloaders: dataloaders(connection), userId: 'admin', connection,
  };

  test('Budget', async () => {
    expect(Budget.id({ _id: 1 })).toBe(1);

    const result = await Budget.account({ source: { name: 'asset' } }, null, { dataloaders: dataloaders(connection), userId });
    expect(result).toMatchSnapshot('account');
  });

  test('Query', async () => {
    const result = await Query.budget(null, {
      dateStart: new Date('2018-05-10'),
      dateEnd: new Date('2018-05-20'),
      account: 'Food',
    }, context);
    expect(result).toMatchSnapshot('budget');
  });

  describe('Mutation', () => {
    test('addBudget', async () => {
      const result = await Mutation.addBudget(null, {
        input: {
          account: 'Food',
          date: '2018-05-10',
          amount: 564,
        },
      }, context);
      expect(result).toMatchSnapshot();
    });

    test('updateBudget', async () => {
      const result = await Mutation.updateBudget(null, {
        id: '2',
        input: {
          amount: 104,
        },
      }, context);
      expect(result).toMatchSnapshot();
    });

    test('deleteBudget', async () => {
      const result = await Mutation.deleteBudget(null, { id: '1' }, context);
      expect(result).toMatchSnapshot();
    });
  });
});
