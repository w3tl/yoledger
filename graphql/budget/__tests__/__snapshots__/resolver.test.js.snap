// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget resolver Budget: account 1`] = `undefined`;

exports[`budget resolver Mutation addBudget 1`] = `
Object {
  "budget": Object {
    "_id": "50",
    "account": "Food",
    "amount": 564,
    "date": 2018-05-10T00:00:00.000Z,
    "userId": "admin",
  },
}
`;

exports[`budget resolver Mutation deleteBudget 1`] = `
Object {
  "success": true,
}
`;

exports[`budget resolver Mutation updateBudget 1`] = `
Object {
  "budget": Object {
    "_id": "2",
    "account": "Food",
    "amount": 55,
    "date": 2018-05-10T00:00:00.000Z,
  },
}
`;

exports[`budget resolver Query: budget 1`] = `
Array [
  Object {
    "_id": "2",
    "account": "Food",
    "amount": 55,
    "date": 2018-05-10T00:00:00.000Z,
  },
]
`;
